File: color.go
package uipack

import (
	"bufio"
	"fmt"
	"strconv"
	"strings"
)

type ColorSpace uint8

const (
	SRGB = iota
	ExtendedSRGB
	DisplayP3
)

// A color value.
type Color struct {
	Red, Green, Blue, Alpha float64
	ColorSpace              ColorSpace
}

var Black = Color{
	Red:        0.0,
	Green:      0.0,
	Blue:       0.0,
	Alpha:      1.0,
	ColorSpace: DisplayP3,
}

func (color *Color) ParseHexString(v string) {
	// Parsing '#RRGGBB' or '#RRGGBBAA'
	v = strings.TrimPrefix(v, "#")
	switch len(v) {
	case 8:
		values, _ := strconv.ParseUint(string(v), 16, 32)
		color.Alpha = float64(uint8(values&0xFF)) / 255.0
		color.Red = float64(uint8((values>>24)&0xFF)) / 255.0
		color.Green = float64(uint8((values>>16)&0xFF)) / 255.0
		color.Blue = float64(uint8((values>>8)&0xFF)) / 255.0
	case 6:
		values, _ := strconv.ParseUint(string(v), 16, 24)
		color.Alpha = 1.0
		color.Red = float64(uint8(values>>16)) / 255.0
		color.Green = float64(uint8((values>>8)&0xFF)) / 255.0
		color.Blue = float64(uint8(values&0xFF)) / 255.0
	default:
		panic("Invalid color format")
	}
}

func (color *Color) ToHexString() string {
	toBytes := func(v float64) byte {
		return byte(uint32(v * 255.0))
	}
	return fmt.Sprintf("%02x%02x%02x%02x", toBytes(color.Alpha), toBytes(color.Red), toBytes(color.Green), toBytes(color.Blue))
}

// Binary encoding

func (color *Color) Encode(writer *bufio.Writer) error {
	err := writer.WriteByte(byte(color.ColorSpace))
	if err != nil {
		return err
	}
	err = writeFloat64(writer, color.Red)
	if err != nil {
		return err
	}
	err = writeFloat64(writer, color.Green)
	if err != nil {
		return err
	}
	err = writeFloat64(writer, color.Blue)
	if err != nil {
		return err
	}
	err = writeFloat64(writer, color.Alpha))
	if err != nil {
		return err
	}

	return nil
}

func (color *Color) Decode(reader *bufio.Reader) error {
	spaceByte, err := reader.ReadByte()
	if err != nil {
		return err
	}
	color.ColorSpace = ColorSpace(spaceByte)

	color.Red, err = readFloat64(reader)
	if err != nil {
		return err
	}
	color.Green, err = readFloat64(reader)
	if err != nil {
		return err
	}
	color.Blue, err = readFloat64(reader)
	if err != nil {
		return err
	}
	color.Alpha, err = readFloat64(reader)
	if err != nil {
		return err
	}
	return nil
}

File: packages/uipack/lib/src/models/color.dart
import '../encoding/binary_encoding.dart';

enum ColorSpace {
  srgb(0),
  extendedSrgb(1),
  displayP3(2);

  const ColorSpace(this.value);
  final int value;

  static ColorSpace fromValue(int value) {
    switch (value) {
      case 0: return ColorSpace.srgb;
      case 1: return ColorSpace.extendedSrgb;
      case 2: return ColorSpace.displayP3;
      default: throw ArgumentError('Invalid ColorSpace value: $value');
    }
  }
}

class Color {
  double red;
  double green;
  double blue;
  double alpha;
  ColorSpace colorSpace;

  Color({
    required this.red,
    required this.green,
    required this.blue,
    required this.alpha,
    required this.colorSpace,
  });

  static final Color black = Color(
    red: 0.0,
    green: 0.0,
    blue: 0.0,
    alpha: 1.0,
    colorSpace: ColorSpace.displayP3,
  );

  void parseHexString(String value) {
    String hex = value.startsWith('#') ? value.substring(1) : value;
    
    switch (hex.length) {
      case 8:
        final intValue = int.parse(hex, radix: 16);
        alpha = (intValue & 0xFF) / 255.0;
        red = ((intValue >> 24) & 0xFF) / 255.0;
        green = ((intValue >> 16) & 0xFF) / 255.0;
        blue = ((intValue >> 8) & 0xFF) / 255.0;
        break;
      case 6:
        final intValue = int.parse(hex, radix: 16);
        alpha = 1.0;
        red = ((intValue >> 16) & 0xFF) / 255.0;
        green = ((intValue >> 8) & 0xFF) / 255.0;
        blue = (intValue & 0xFF) / 255.0;
        break;
      default:
        throw ArgumentError('Invalid color format: $value');
    }
  }

  String toHexString() {
    int toBytes(double value) => (value * 255.0).round() & 0xFF;
    return '${toBytes(alpha).toRadixString(16).padLeft(2, '0')}'
           '${toBytes(red).toRadixString(16).padLeft(2, '0')}'
           '${toBytes(green).toRadixString(16).padLeft(2, '0')}'
           '${toBytes(blue).toRadixString(16).padLeft(2, '0')}';
  }

  void encode(ByteDataWriter writer) {
    writer.writeUint8(colorSpace.value);
    writer.writeFloat64(red);
    writer.writeFloat64(green);
    writer.writeFloat64(blue);
    writer.writeFloat64(alpha);
  }

  void decode(ByteDataReader reader) {
    colorSpace = ColorSpace.fromValue(reader.readUint8());
    red = reader.readFloat64();
    green = reader.readFloat64();
    blue = reader.readFloat64();
    alpha = reader.readFloat64();
  }

  @override
  String toString() => '#${toHexString()}';
}